bento/src/main/java/com/yelp/android/bento/core/AsyncInflationBridge.kt

```diff
@@ -0,0 +1,91 @@
+package com.yelp.android.bento.core
+
+import android.view.View
+import androidx.recyclerview.widget.RecyclerView
+import com.yelp.android.bento.componentcontrollers.RecyclerViewComponentController.constructViewHolder
+import kotlinx.coroutines.*
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import java.util.concurrent.Executors
+
+
+/**
+ * This acts as a bridge between RecyclerViewComponentController and the underlying
+ * RecyclerView.Adapter. It intercepts the addComponent() calls and inflates views with
+ * [BentoAsyncLayoutInflater]. Components are added to the RecyclerView when the inflation
+ * is finished.
+ */
+internal class AsyncInflationBridge(val recyclerView: RecyclerView) {
+
+    private companion object {
+        private val lock = Mutex()
+        private val dispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+    }
+
+    private val viewHolderMap = mutableMapOf<Class<out ComponentViewHolder<*, *>>,
+            MutableList<ComponentViewHolder<*, *>>>()
+    private val viewMap = mutableMapOf<ComponentViewHolder<*, *>, View?>()
+
+    /**
+     * Inflates the views and creates the view holder for [component] on a background thread,
+     * getting them nice and ready for RecyclerViewComponentController to use them. The
+     */
+    fun asyncInflateViewsForComponent(component: Component, addComponentCallback: () -> Unit) {
+        GlobalScope.launch(dispatcher) {
+            val deferredInflations = (0 until component.count).map { i ->
+                async {
+                    val viewHolderType = component.getHolderType(i)
+                    val viewHolder: ComponentViewHolder<*, *> = constructViewHolder(viewHolderType)
+                    addViewHolder(viewHolder, viewHolderType)
+                    val pair = BentoAsyncLayoutInflater.inflate(viewHolder, recyclerView)
+                    viewMap[viewHolder] = pair.second
+                }
+            }
+            lock.withLock {
+                deferredInflations.awaitAll()
+            }
+            withContext(Dispatchers.Main) {
+                try {
+                    addComponentCallback()
+                } catch (e: IllegalArgumentException) {
+                    // TODO: handle crash if component is already added.
+                }
+            }
+        }
+    }
+
+    /**
+     * Retrieves a previously async inflated view if there is one.
+     */
+    @Synchronized
```

Comment:
This is required to pull views in order from the main thread. We could remove this and make `viewMap` the concurrent version instead, if you like?

Reply:
I still cannot understand why it would help, or do anything. We only synchronize `getView`, but getView is only called from the `RecyclerViewComponentController` class in the `onCreateViewHolder`. And I think this gets called by the RecyclerView and the GapWorker, but on the main thread. So I'm lost.

I'm scratching my head but I don't get it. I would feel more confident with a concurrent map I think, unless I finally understand why this work, or you can explain it.

Reply:
As we agreed on Slack, although there isn't much risk of a concurrent mishap for that map, it's still safer to use a `ConcurrentHashMap`. 👍